ext/drx/drx.c

```diff
@@ -2429,3 +2441,363 @@ drx_expand_scatter_gather(void *drcontext, instrlist_t *bb, OUT bool *expanded)
+{
+    if (*allow_for_unknown_instr_count++ < DRX_RESTORE_EVENT_ALLOW_FOR_UNKNOWN_INSTR)
+        return true;
+    return false;
+}
+
+static void
+advance_state(int *detect_state, int new_detect_state, int *allow_for_unknown_instr_count)
+{
+    *detect_state = new_detect_state;
+    *allow_for_unknown_instr_count = 0;
+}
+
+static void
+drx_try_to_detect_avx512_gather_sequence(void *drcontext, dr_restore_state_info_t *info,
+                                         instr_t *inst, scatter_gather_info_t *sg_info)
+{
+    byte *prev_pc, *pc;
+    byte *restore_dest_mask_start = NULL;
+    byte *restore_scratch_mask_start = NULL;
+    int detect_state = 0;
+    int allow_for_unknown_instr_count = 0;
+    reg_id_t tmp_xmm = DR_REG_NULL;
+    reg_id_t gpr_bit_mask = DR_REG_NULL;
+    reg_id_t gpr_save_scratch_mask = DR_REG_NULL;
+    uint scalar_mask_update_no = 0;
+    pc = info->fragment_info.cache_start_pc;
+    /* As the state machine is looking for blocks of code that the fault may hit, the 128
+     * bytes is a conservative approximation of the block's size, see (a) and (b) above.
+     */
+    while (pc <= info->raw_mcontext->pc + 128) {
+        instr_reset(drcontext, inst);
+        prev_pc = pc;
+        pc = decode(drcontext, pc, inst);
+        if (pc == NULL) {
+            /* Upon a decoding error we simply give up. */
+            return;
+        }
+        /* If there is a gather or scatter instruction in the code cache, then it is wise
+         * to assume that this is not an emulated sequence that we need to examine
+         * further.
+         */
+        if (instr_is_gather(inst))
+            return;
+        if (instr_is_scatter(inst))
+            return;
+        switch (detect_state) {
+        case DRX_DETECT_RESTORE_AVX512_GATHER_EVENT_STATE_0:
+            if (instr_reads_memory(inst)) {
+                opnd_t dst0 = instr_get_dst(inst, 0);
+                if (opnd_is_reg(dst0) && reg_is_gpr(opnd_get_reg(dst0))) {
+                    restore_dest_mask_start = pc;
+                    advance_state(&detect_state,
+                                  DRX_DETECT_RESTORE_AVX512_GATHER_EVENT_STATE_1,
+                                  &allow_for_unknown_instr_count);
+                    break;
+                }
+            }
+            break;
+        case DRX_DETECT_RESTORE_AVX512_GATHER_EVENT_STATE_1:
+            if (instr_get_opcode(inst) == OP_vextracti32x4) {
+                reg_id_t tmp_reg = opnd_get_reg(instr_get_dst(inst, 0));
+                if (!reg_is_strictly_xmm(tmp_reg))
+                    break;
+                tmp_xmm = tmp_reg;
+                advance_state(&detect_state,
+                              DRX_DETECT_RESTORE_AVX512_GATHER_EVENT_STATE_2,
+                              &allow_for_unknown_instr_count);
+                break;
+            }
+            /* Intentionally not else if */
+            if (allow_for_unknown_instr_inc(&allow_for_unknown_instr_count))
```

Comment:
Why is this unknown+reset-to-0 at the end of every state except state 0?   That looks like a bug: if I'm in state 0 and never find a load I will never hit the unknown instr limit.

Reply:
Because in state 0, if we haven't seen the beginning of the sequence, there is no point in ignoring instructions. We are already in STATE_0 anyway. I have added a comment. Resolving, please re-open if necessary.